File: rcc/clients/tcp/openrgb/openrgb_client.py
import os
import re
import socket
import subprocess
from typing import Optional
import threading
import time

# pylint:disable=E0611, E0401
from openrgb import OpenRGBClient
from openrgb.orgb import Device
from openrgb.utils import RGBColor

from rcc.models.rgb_brightness import RgbBrightness
from rcc.models.usb_identifier import UsbIdentifier
from rcc.utils.constants import orgb_path, udev_path
from rcc.utils.event_bus import event_bus
from rcc.utils.logger import Logger
from rcc.utils.singleton import singleton
from rcc.clients.tcp.openrgb.effects.base.abstract_effect import AbstractEffect
from rcc.clients.tcp.openrgb.effects.breathing import breathing_effect
from rcc.clients.tcp.openrgb.effects.dance_floor import dance_floor
from rcc.clients.tcp.openrgb.effects.digital_rain import digital_rain
from rcc.clients.tcp.openrgb.effects.rain import rain
from rcc.clients.tcp.openrgb.effects.rainbow_wave import rainbow_wave
from rcc.clients.tcp.openrgb.effects.spectrum_cycle import spectrum_cycle
from rcc.clients.tcp.openrgb.effects.starry_night import starry_night
from rcc.clients.tcp.openrgb.effects.static import static_effect


@singleton
class OpenRgbClient:
    """Client to communicate with OpenRGB Server"""

    def __init__(self):
        self.logger = Logger()
        event_bus.on("stop", self.stop)
        self._available_devices: list[Device] = []
        self._client = None
        self._orgb_thread = None
        self._orgb_process = None
        self._port = 6472
        self._compatible_devices = None
        self._available_effects: list[AbstractEffect] = [
            breathing_effect,
            dance_floor,
            digital_rain,
            rain,
            rainbow_wave,
            spectrum_cycle,
            starry_night,
            static_effect,
        ]
        self.start()

    def get_compatible_devices(self) -> list[UsbIdentifier] | None:
        """Get list of compatible devices"""
        return self._compatible_devices

    def start(self):
        """Initialize server and client"""
        self.logger.info("Initializing OpenRgbClient")
        subprocess.run(
            ["asusctl", "led-mode", "static", "-c", "000000"],
            check=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            encoding="utf-8",
        )
        self.logger.add_tab()
        self._start_orgb_process()
        self._find_compatible_devices()
        self._start_client()
        self._get_available_devices()
        self._client.set_color(RGBColor(0, 0, 0), True)
        self.logger.rem_tab()

    def stop(self, _=None):
        """Stop server and client"""
        self.logger.info("Stopping OpenRgbClient")
        self.logger.add_tab()
        for mode in self._available_effects:
            mode.stop()
        self._client.set_color(RGBColor(0, 0, 0), True)
        time.sleep(0.2)
        self._stop_orgb_process()
        self.logger.rem_tab()

    def _start_client(self):
        """Initialize client"""
        self.logger.info("Connecting to server")
        self.logger.add_tab()
        self._client = OpenRGBClient(port=self._port, name="RogControlCenter")
        self._client.connect()
        self.logger.info("Connected")
        self.logger.rem_tab()

    def _stop_client(self):
        """Stop client"""
        self._client.disconnect()
        self._client = None

    def _get_available_devices(self):
        self.logger.info("Getting available devices")
        self.logger.add_tab()
        self._available_devices = self._client.ee_devices
        for dev in self._available_devices:
            self.logger.info(dev.name)
        self.logger.rem_tab()

    def _start_orgb_process(self):
        self.logger.info("Starting OpenRgb server")
        self.logger.add_tab()
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.bind(("", 0))
            self._port = s.getsockname()[1]

        self._orgb_thread = threading.Thread(name="ORGBServer", target=self._run_in_background)
        self._orgb_thread.start()

        self._wait_for_server()
        self.logger.info("OpenRgb server ready")
        self.logger.rem_tab()

    @property
    def compatible_devices(self) -> list[UsbIdentifier]:
        """Getter for compatible devices"""
        return self._compatible_devices

    def _stop_orgb_process(self):
        self._orgb_process.kill()

    def _run_in_background(self):
        self._orgb_process = subprocess.Popen(  # pylint: disable=R1732
            [orgb_path, "--server-host", "localhost", "--server-port", str(self._port)],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
        )  # pylint: disable=R1732

        for _ in self._orgb_process.stdout:
            pass

        for _ in self._orgb_process.stderr:
            pass

        self._orgb_process.wait()
        self.logger.info(f"OpenRgb finished with code {self._orgb_process.returncode}")
        self._orgb_process = None

    def _wait_for_server(self) -> bool:
        while True:
            try:
                with socket.create_connection(("localhost", self._port), timeout=1):
                    break
            except Exception:
                time.sleep(0.1)

    def _find_compatible_devices(self):
        self.logger.info("Reading udev rules")
        compatible_devices = None

        if compatible_devices is None:
            if os.path.exists(udev_path):
                with open(udev_path, "r") as file:
                    content = file.read()
                lines = content.split("\n")

                regex = r'SUBSYSTEMS==".*?", ATTRS{idVendor}=="([0-9a-fA-F]+)", ATTRS{idProduct}=="([0-9a-fA-F]+)"'
                results = []
                for line in lines:
                    match = re.search(regex, line)
                    if match:
                        results.append(UsbIdentifier(match.group(1), match.group(2)))
                compatible_devices = results

        self._compatible_devices = compatible_devices

    def get_available_effects(self) -> list[AbstractEffect]:
        """Get all effects"""
        return [effect.name for effect in self._available_effects]

    def supports_color(self, effect) -> bool:
        """Check if effect supports color"""
        effects = [e for e in self._available_effects if e.name == effect]
        if len(effects) > 0:
            return effects[0].supports_color
        return False

    def get_color(self, effect) -> str | None:
        """Get current hex color or none"""
        effects = [e for e in self._available_effects if e.name == effect]
        if len(effects) > 0:
            return effects[0].color
        return None

    def apply_effect(self, effect: str, brightness: RgbBrightness, color: Optional[str] = None) -> None:
        """Apply effect with specified brightness and color"""
        inst = [i for i in self._available_effects if i.name == effect]
        if inst:
            for mode in self._available_effects:
                mode.stop()

            inst[0].start(self._available_devices, brightness, RGBColor.fromHEX(color or "#000000"))


open_rgb_client = OpenRgbClient()

File: RogControlCenter/src/clients/tcp/open_rgb/open_rgb_client.cpp
#include "../../../../include/clients/tcp/open_rgb/open_rgb_client.hpp"

#include <string>
#include <vector>

#include "../../../../include/clients/shell/asusctl_client.hpp"
#include "../../../../include/clients/tcp/open_rgb/effects/breathing_effect.hpp"
#include "../../../../include/clients/tcp/open_rgb/effects/dance_floor_effect.hpp"
#include "../../../../include/clients/tcp/open_rgb/effects/digital_rain_effect.hpp"
#include "../../../../include/clients/tcp/open_rgb/effects/drops_effect.hpp"
#include "../../../../include/clients/tcp/open_rgb/effects/gaming.hpp"
#include "../../../../include/clients/tcp/open_rgb/effects/rainbow_wave.hpp"
#include "../../../../include/clients/tcp/open_rgb/effects/spectrum_cycle_effect.hpp"
#include "../../../../include/clients/tcp/open_rgb/effects/starry_night_effect.hpp"
#include "../../../../include/clients/tcp/open_rgb/effects/static_effect.hpp"
#include "../../../../include/events/event_bus.hpp"
#include "../../../../include/models/hardware/rgb_brightness.hpp"
#include "../../../../include/utils/file_utils.hpp"
#include "../../../../include/utils/net_utils.hpp"
#include "../../../../include/utils/process_utils.hpp"

void OpenRgbClient::initialize() {
	if (!FileUtils::exists(Constants::UDEV_RULES) || !md5SumClient.available() ||
		md5SumClient.getChecksum(Constants::ORGB_UDEV_PATH) != md5SumClient.getChecksum(Constants::Constants::UDEV_RULES)) {
		logger.info("Configuring UDEV rules");
		Logger::add_tab();

		shell.run_command("cp " + Constants::ORGB_UDEV_PATH + " " + Constants::TMP_UDEV_PATH);
		shell.run_elevated_command("mv " + Constants::TMP_UDEV_PATH + " " + Constants::UDEV_RULES + " && chmod 777 " + Constants::UDEV_RULES +
								   " && udevadm control --reload-rules && udevadm trigger");

		Logger::rem_tab();
	}

	availableEffects.push_back(std::unique_ptr<AbstractEffect>(&BreathingEffect::init(client)));
	availableEffects.push_back(std::unique_ptr<AbstractEffect>(&DanceFloorEffect::init(client)));
	availableEffects.push_back(std::unique_ptr<AbstractEffect>(&DigitalRainEffect::init(client)));
	availableEffects.push_back(std::unique_ptr<AbstractEffect>(&DropsEffect::init(client)));
	availableEffects.push_back(std::unique_ptr<AbstractEffect>(&GamingEffect::init(client)));
	availableEffects.push_back(std::unique_ptr<AbstractEffect>(&RainbowWave::init(client)));
	availableEffects.push_back(std::unique_ptr<AbstractEffect>(&SpectrumCycleEffect::init(client)));
	availableEffects.push_back(std::unique_ptr<AbstractEffect>(&StarryNightEffect::init(client)));
	availableEffects.push_back(std::unique_ptr<AbstractEffect>(&StaticEffect::init(client)));

	eventBus.onApplicationStop([this]() {
		stop();
	});

	start();
}

void OpenRgbClient::configureUdev() {
}

void OpenRgbClient::start() {
	logger.info("Starting OpenRgbClient");
	Logger::add_tab();

	if (asusCtlClient.available()) {
		asusCtlClient.turnOffAura();
	}
	startOpenRgbProcess();
	startOpenRgbClient();
	getAvailableDevices();
	Logger::rem_tab();
}

void OpenRgbClient::stop() {
	logger.info("Stopping OpenRgbClient");
	Logger::add_tab();
	for (auto& effect : availableEffects) {
		effect->stop();
	}
	for (auto& dev : detectedDevices) {
		client.setDeviceColor(dev, Color::Black);
	}
	client.disconnect();
	ProcessUtils::sleep(100);
	stopOpenRgbProcess();
	Logger::rem_tab();
}

const CompatibleDeviceArray OpenRgbClient::getCompatibleDevices() {
	return compatibleDevices;
}

void OpenRgbClient::startOpenRgbProcess() {
	logger.info("Starting OpenRGB server");
	Logger::add_tab();
	port		 = NetUtils::getRandomFreePort();
	runnerThread = std::thread(&OpenRgbClient::runner, this);
	while (NetUtils::isPortFree(port)) {
		ProcessUtils::sleep(50);
	}
	logger.info("OpenRgb server ready");
	Logger::rem_tab();
}

void OpenRgbClient::stopOpenRgbProcess() {
	logger.info("Stopping OpenRGB server");
	Logger::add_tab();
	ProcessUtils::sendSignal(pid, SIGKILL);
	if (runnerThread.joinable()) {
		runnerThread.join();
	}
	logger.info("OpenRGB server killed");
	Logger::rem_tab();
}

void OpenRgbClient::runner() {
	std::vector<std::string> argsStr = {"--server-host", "localhost", "--server-port", std::to_string(port), "-v"};
	std::vector<char*> argv;
	argv.push_back(const_cast<char*>(Constants::ORGB_PATH.c_str()));
	for (auto& s : argsStr) {
		argv.push_back(const_cast<char*>(s.c_str()));
	}
	argv.push_back(nullptr);

	std::vector<std::string> envStrings = shell.copyEnviron();
	envStrings.push_back("LD_LIBRARY_PATH=");
	std::vector<char*> env;
	for (auto& s : envStrings) {
		env.push_back(s.data());
	}
	env.push_back(nullptr);

	pid			  = shell.launch_process(Constants::ORGB_PATH.c_str(), argv.data(), env.data(),
										 Constants::LOG_DIR + "/" + Constants::LOG_ORGB_FILE_NAME + ".log");
	int exit_code = shell.wait_for(pid);
	logger.info("Command finished with exit code {}", exit_code);
}

void OpenRgbClient::startOpenRgbClient() {
	logger.info("Connecting to server");
	Logger::add_tab();
	client.connect("localhost", port);
	logger.info("Connected");
	Logger::rem_tab();
}

void OpenRgbClient::getAvailableDevices() {
	logger.info("Getting available devices");
	Logger::add_tab();

	detectedDevices = client.requestDeviceList().devices;
	for (auto& dev : detectedDevices) {
		const orgb::Mode* directMode = dev.findMode("Direct");
		if (directMode) {
			logger.info("{}", dev.name);
			client.changeMode(dev, *directMode);
			client.setDeviceColor(dev, orgb::Color::Black);
		}
	}

	Logger::rem_tab();
}
const std::vector<std::string> OpenRgbClient::getAvailableEffects() {
	std::vector<std::string> result;
	for (auto& effect : availableEffects) {
		result.push_back(effect->getName());
	}
	return result;
}

void OpenRgbClient::applyEffect(const std::string& effectName, const RgbBrightness& brightness) {
	for (const auto& effect : availableEffects) {
		effect->stop();
	}
	for (const auto& effect : availableEffects) {
		if (effect->getName() == effectName) {
			effect->start(detectedDevices, brightness);
			break;
		}
	}
}

void OpenRgbClient::disableDevice(const std::string& devName) {
	for (auto& dev : detectedDevices.devices()) {
		if (dev->name == devName) {
			dev->enabled = false;
		}
	}
}